Update Project_Planning

connorholley opened the pull request:
Accidentally merged my pull request so more punctuation for you to review again! :joy: I have a rough draft of some code to access the csv file and a general description of my naïve function.

connorholley committed:
Update Project_Planning

sean-chester reviewed (approved):
Thanks @connorholley . To summarise our meeting yesterday:

 * See if you can implement the jaccard similarity using a call to `std::set_intersection<>`
    - This will likely require learning about comparators, function overloads, lambdas, and/or functors.
    - You might also want to consider whether it makes sense to define a geospatial object/struct to keep track of your timestamp + long/lat. Alternatively, you may find `std::tuple<>` to be more useful here than just `std::vector<>`

connorholley committed:
Update Baboon_Reader.cpp

sean-chester commented:
Hey @connorholley , some quick `git` tips:
 1. After a pull request (PR) is approved, it is a good idea [to merge the pull request](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request). This is, in fact, what the name is referring to, that somebody is making a request to pull a change set into an existing branch/repo. This way, the `main` branch always has the latest approved changes.
 2. When creating a new change set (e.g., here in commit 000c133), it is a good idea [to open a new branch](https://stackoverflow.com/q/3899627/2769271). If you do that, then you can open up a new PR that is specific to this change set. That has two advantages: a) it avoids the confusion of having your reviewer look at a PR that they already reviewed; and b) it keeps change sets self-contained which makes it easier to keep track of what's happening.

So, see if you can incorporate these two ideas with regards to this pull request (after I re-review it) and the  next change set that you have planned.

sean-chester commented on Baboon_stuff/Baboon_Reader.cpp:
```diff
@@ -1,77 +1,152 @@
-      vector<string> result;
-
-      while(s_stream.good()) //goes through each line (while there are lines to go through) and seperates based on the delimeter (the comma)
+class Baboon_Reader {
```
Nice to see the use of classes now. :+1:

sean-chester commented on Baboon_stuff/Baboon_Reader.cpp:
```diff
@@ -1,77 +1,152 @@
-      while(s_stream.good()) //goes through each line (while there are lines to go through) and seperates based on the delimeter (the comma)
+class Baboon_Reader {
+public:
+  map<string, vector<tuple<double,double,double>>> readFile(){
```
You may want to consider the syntax for creating aliases:
```c++
using myname = std::map< std::string, std::vector< std::tuple< double, double, double > > >;
```
Then elsewhere in your code you can use the better named alias than this really long compound data type.

connorholley replied:
This is very helpful, I implemented it this morning, makes the code much more attractive. Thank you

sean-chester commented on Baboon_stuff/Baboon_Reader.cpp:
```diff
@@ -1,77 +1,152 @@
+            getline(s_stream, substr, ',');
+            result.push_back(substr);
+          }
+          if (find(baboon_names.begin(), baboon_names.end(),result[8]) != baboon_names.end()){
```
It's good practice to avoid using raw numbers like this, as it can often be hard to know where that "magic number" came from. Try, for example,
```c++
auto const name_index = 8u;
if( find( baboon_names.cbegin(), baboon_names.cend(), result[ name_index ] ) != baboon_names.cend() ) // stuff
```

connorholley replied:
Another helpful formatting recommendation! Will look up the documentation to find out exactly what is happening here

sean-chester commented on Baboon_stuff/Baboon_Reader.cpp:
```diff
@@ -1,77 +1,152 @@
+    for (int i = 0; i < dicty[baboon_name].size(); i++) {
+      for(int j = 0; j< baboon_names.size();j++){
+        double totalIntersections = 0;
+        for(int k =0; k< dicty[baboon_names[j]].size();k++){
+          if (get<1>(dicty[baboon_name][i])-longitude_tolerance <= get<1>(dicty[baboon_names[j]][k]) &&get<1>(dicty[baboon_names[j]][k])<=get<1>(dicty[baboon_name][i])+longitude_tolerance){
+            if (get<2>(dicty[baboon_name][i])-latitude_tolerance <= get<2>(dicty[baboon_names[j]][k]) &&get<2>(dicty[baboon_names[j]][k])<=get<2>(dicty[baboon_name][i])+latitude_tolerance){
```
This is a great example of where you might benefit from some _refactoring_: can you transform these three lines into just one function call?

connorholley replied:
I know I can combine the two if statements as with && it will only evaluate the right side if the left side is true. With regards to transforming the lines into a function call I assume you mean taking the mess of nested for loops and declaring them as their own function and then just calling it here? Or is there some better function used in c++ that I am missing? I also just recently watched a video on list comprehension, it seems to make the syntax more concise. Would that qualify as a form of refactoring? If so I can work on implementing that in place of these burdensome for loops.

sean-chester reviewed (commented):
@connorholley , I'm not sure whether you wanted me to review these changes yet or whether you just happened to push to the same branch for which you already had an open PR, but I added a few small comments that I hope might help.

connorholley committed:
Update Baboon_Reader.cpp

Made the recommended changes using aliases. Cleaned up magic numbers by using more descriptive variable names
